{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Systems genetics 2020\n",
    "## Final Project  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "liver_exp = pd.read_csv('data/GSE17522_series_matrix_liver.txt', sep = \"\\t\")\n",
    "brain_exp = pd.read_csv('data/GSE36674_series_matrix_hypothalamus.txt', sep = \"\\t\")\n",
    "liver_annotations = pd.read_csv('data/annotations_liver_GPL6466-9752.txt', sep = \"\\t\")\n",
    "genotypes = pd.read_excel('data/genotypes.xls', headers=None)\n",
    "pheotypes = pd.read_excel('data/phenotypes.xls')\n",
    "mgi = pd.read_csv('data/MGI_Coordinates.Build37.rpt.txt', sep = \"\\t\", error_bad_lines=False, warn_bad_lines=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Gene expression data preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### • Make sure that the data is normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Data were extracted from the scanned image using Agilent Feature Extraction software version 6.1. A total of 122 arrays were run in 8 batches. The samples were semirandomly distributed throughout the batches prior to microarray analysis in order to separate sexes and strains, and to minimize between- and within- batch bias. Technical and biological replicates were run both within each batch and between batches. The microarray data was deposited in the UNC Microarray Database and extracted using Log2 ratios of the mean red channel intensity over the mean green channel intensity. This was followed by LOWESS normalization to remove the intensity dependent dye bias3. Neither the genes nor the arrays were centered. Inter-batch normalization was carried out using a nested ANOVA mixed model with samples within each batch crossed with sex and strain.'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "liver_exp[liver_exp['!Sample_title'] == '!Sample_data_processing']['Liver_C57BL6J_M_B1_rep1'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Probe intensity values were extracted using the Affymetrix GeneChip Operating Software and RMA normalized'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "brain_exp[brain_exp['!Sample_title'] == '!Sample_data_processing']['BXD44_F'].iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### • Merge data file with annotation file to get your input matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove metadata rows to get the raw data only\n",
    "liver_exp = liver_exp[~liver_exp['!Sample_title'].str.contains('!Sample_', na=False)]\n",
    "liver_exp = liver_exp[~liver_exp['!Sample_title'].str.contains('!series_', na=False)]\n",
    "liver_exp = liver_exp[liver_exp['!Sample_title'] != 'ID_REF']\n",
    "\n",
    "# Rename ID column to match the annotation matrix\n",
    "liver_exp = liver_exp.rename(columns = {'!Sample_title' : 'ID'})\n",
    "\n",
    "# Merge with annotation matrix to get the gene identifier (gene symbol)\n",
    "id_cols = ['ID', 'GENE_SYMBOL']\n",
    "liver_annotations = liver_annotations[id_cols]\n",
    "liver_matrix = liver_annotations.merge(liver_exp, left_on='ID', right_on='ID')\n",
    "\n",
    "# Keep BXD columns and identifier columns only\n",
    "liver_bxd_cols = list(liver_matrix.filter(regex=(\"BXD*\")).columns)\n",
    "liver_matrix = liver_matrix[id_cols + liver_bxd_cols]\n",
    "\n",
    "# Rename BXD columns\n",
    "liver_matrix = liver_matrix.rename(columns={col : col.split('_')[1] for col in liver_matrix.columns.drop(id_cols)})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### • Remove rows with no gene identifier, <br> • Remove rows with low maximal value.  <br> • Remove rows with low variance.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>BXD1</th>\n",
       "      <th>BXD12</th>\n",
       "      <th>BXD23</th>\n",
       "      <th>BXD34</th>\n",
       "      <th>BXD28</th>\n",
       "      <th>BXD29</th>\n",
       "      <th>BXD39</th>\n",
       "      <th>BXD40</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD39</th>\n",
       "      <th>BXD86</th>\n",
       "      <th>BXD2</th>\n",
       "      <th>BXD69</th>\n",
       "      <th>BXD73</th>\n",
       "      <th>BXD92</th>\n",
       "      <th>BXD51</th>\n",
       "      <th>BXD48</th>\n",
       "      <th>BXD85</th>\n",
       "      <th>BXD45</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_51_P100021</td>\n",
       "      <td>Hivep3</td>\n",
       "      <td>-0.011</td>\n",
       "      <td>-0.075</td>\n",
       "      <td>-0.091</td>\n",
       "      <td>-0.148</td>\n",
       "      <td>0.052</td>\n",
       "      <td>0.06</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>-0.047</td>\n",
       "      <td>...</td>\n",
       "      <td>0.061</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>0.239</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>0.305</td>\n",
       "      <td>0.139</td>\n",
       "      <td>0.053</td>\n",
       "      <td>0.099</td>\n",
       "      <td>0.044</td>\n",
       "      <td>0.137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_51_P100034</td>\n",
       "      <td>Mif4gd</td>\n",
       "      <td>-0.641</td>\n",
       "      <td>-0.396</td>\n",
       "      <td>0.399</td>\n",
       "      <td>-0.892</td>\n",
       "      <td>-0.326</td>\n",
       "      <td>0.002</td>\n",
       "      <td>-0.406</td>\n",
       "      <td>-0.642</td>\n",
       "      <td>...</td>\n",
       "      <td>0.146</td>\n",
       "      <td>-0.216</td>\n",
       "      <td>-0.419</td>\n",
       "      <td>-0.275</td>\n",
       "      <td>-0.441</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.187</td>\n",
       "      <td>-0.234</td>\n",
       "      <td>-0.538</td>\n",
       "      <td>-0.545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_51_P100052</td>\n",
       "      <td>Slitrk2</td>\n",
       "      <td>0.192</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.789</td>\n",
       "      <td>0.583</td>\n",
       "      <td>0.584</td>\n",
       "      <td>0.663</td>\n",
       "      <td>0.515</td>\n",
       "      <td>0.391</td>\n",
       "      <td>...</td>\n",
       "      <td>0.414</td>\n",
       "      <td>0.234</td>\n",
       "      <td>0.43</td>\n",
       "      <td>0.233</td>\n",
       "      <td>0.612</td>\n",
       "      <td>0.087</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.208</td>\n",
       "      <td>0.247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_51_P100065</td>\n",
       "      <td>Lnx1</td>\n",
       "      <td>-1.589</td>\n",
       "      <td>-1.777</td>\n",
       "      <td>-1.648</td>\n",
       "      <td>-1.68</td>\n",
       "      <td>-1.834</td>\n",
       "      <td>-1.608</td>\n",
       "      <td>-1.816</td>\n",
       "      <td>-1.021</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.023</td>\n",
       "      <td>-2.182</td>\n",
       "      <td>-2.522</td>\n",
       "      <td>-2.401</td>\n",
       "      <td>-1.741</td>\n",
       "      <td>-1.91</td>\n",
       "      <td>-2.115</td>\n",
       "      <td>-1.877</td>\n",
       "      <td>-2.316</td>\n",
       "      <td>-2.343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A_51_P100099</td>\n",
       "      <td>2610002J23Rik</td>\n",
       "      <td>-1.19</td>\n",
       "      <td>-1.013</td>\n",
       "      <td>-1.142</td>\n",
       "      <td>-1.312</td>\n",
       "      <td>-1.163</td>\n",
       "      <td>-0.983</td>\n",
       "      <td>-1.31</td>\n",
       "      <td>-1.042</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.158</td>\n",
       "      <td>-1.209</td>\n",
       "      <td>-1.208</td>\n",
       "      <td>-0.784</td>\n",
       "      <td>-0.869</td>\n",
       "      <td>-0.848</td>\n",
       "      <td>-1.11</td>\n",
       "      <td>-1.159</td>\n",
       "      <td>-0.857</td>\n",
       "      <td>-1.036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20863</th>\n",
       "      <td>A_51_P521307</td>\n",
       "      <td>Olfr1512</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.245</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.334</td>\n",
       "      <td>0.119</td>\n",
       "      <td>0.456</td>\n",
       "      <td>0.595</td>\n",
       "      <td>0.473</td>\n",
       "      <td>...</td>\n",
       "      <td>0.185</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.231</td>\n",
       "      <td>0.359</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.227</td>\n",
       "      <td>0.292</td>\n",
       "      <td>0.026</td>\n",
       "      <td>-0.145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20864</th>\n",
       "      <td>A_51_P521346</td>\n",
       "      <td>C7</td>\n",
       "      <td>1.101</td>\n",
       "      <td>0.931</td>\n",
       "      <td>0.724</td>\n",
       "      <td>1.054</td>\n",
       "      <td>0.896</td>\n",
       "      <td>0.876</td>\n",
       "      <td>1.064</td>\n",
       "      <td>0.986</td>\n",
       "      <td>...</td>\n",
       "      <td>0.841</td>\n",
       "      <td>0.758</td>\n",
       "      <td>0.831</td>\n",
       "      <td>1.14</td>\n",
       "      <td>0.618</td>\n",
       "      <td>0.679</td>\n",
       "      <td>0.823</td>\n",
       "      <td>0.678</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20865</th>\n",
       "      <td>A_51_P521364</td>\n",
       "      <td>Emx2</td>\n",
       "      <td>0.129</td>\n",
       "      <td>0.353</td>\n",
       "      <td>0.373</td>\n",
       "      <td>0.496</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.204</td>\n",
       "      <td>0.475</td>\n",
       "      <td>0.59</td>\n",
       "      <td>...</td>\n",
       "      <td>0.306</td>\n",
       "      <td>0.249</td>\n",
       "      <td>0.347</td>\n",
       "      <td>0.517</td>\n",
       "      <td>0.207</td>\n",
       "      <td>0.322</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.246</td>\n",
       "      <td>0.301</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20866</th>\n",
       "      <td>A_51_P521374</td>\n",
       "      <td>Nudt13</td>\n",
       "      <td>-0.426</td>\n",
       "      <td>-0.363</td>\n",
       "      <td>-0.511</td>\n",
       "      <td>-0.423</td>\n",
       "      <td>-0.414</td>\n",
       "      <td>0.033</td>\n",
       "      <td>-0.047</td>\n",
       "      <td>-0.357</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.108</td>\n",
       "      <td>0.028</td>\n",
       "      <td>-0.146</td>\n",
       "      <td>-0.216</td>\n",
       "      <td>-0.273</td>\n",
       "      <td>-0.409</td>\n",
       "      <td>-0.327</td>\n",
       "      <td>-0.215</td>\n",
       "      <td>-0.505</td>\n",
       "      <td>-0.261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20867</th>\n",
       "      <td>A_51_P522004</td>\n",
       "      <td>Dst</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.498</td>\n",
       "      <td>0.283</td>\n",
       "      <td>0.271</td>\n",
       "      <td>0.676</td>\n",
       "      <td>0.201</td>\n",
       "      <td>0.336</td>\n",
       "      <td>0.348</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.169</td>\n",
       "      <td>0.102</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0.096</td>\n",
       "      <td>0.255</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.045</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.426</td>\n",
       "      <td>0.202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19272 rows × 82 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ID    GENE_SYMBOL    BXD1   BXD12   BXD23   BXD34   BXD28  \\\n",
       "0      A_51_P100021         Hivep3  -0.011  -0.075  -0.091  -0.148   0.052   \n",
       "1      A_51_P100034         Mif4gd  -0.641  -0.396   0.399  -0.892  -0.326   \n",
       "2      A_51_P100052        Slitrk2   0.192   0.367   0.789   0.583   0.584   \n",
       "3      A_51_P100065           Lnx1  -1.589  -1.777  -1.648   -1.68  -1.834   \n",
       "5      A_51_P100099  2610002J23Rik   -1.19  -1.013  -1.142  -1.312  -1.163   \n",
       "...             ...            ...     ...     ...     ...     ...     ...   \n",
       "20863  A_51_P521307       Olfr1512   0.027   0.245    0.28   0.334   0.119   \n",
       "20864  A_51_P521346             C7   1.101   0.931   0.724   1.054   0.896   \n",
       "20865  A_51_P521364           Emx2   0.129   0.353   0.373   0.496   0.328   \n",
       "20866  A_51_P521374         Nudt13  -0.426  -0.363  -0.511  -0.423  -0.414   \n",
       "20867  A_51_P522004            Dst   0.326   0.498   0.283   0.271   0.676   \n",
       "\n",
       "        BXD29   BXD39   BXD40  ...   BXD39   BXD86    BXD2   BXD69   BXD73  \\\n",
       "0        0.06   -0.09  -0.047  ...   0.061   -0.21   0.239   -0.06   0.305   \n",
       "1       0.002  -0.406  -0.642  ...   0.146  -0.216  -0.419  -0.275  -0.441   \n",
       "2       0.663   0.515   0.391  ...   0.414   0.234    0.43   0.233   0.612   \n",
       "3      -1.608  -1.816  -1.021  ...  -2.023  -2.182  -2.522  -2.401  -1.741   \n",
       "5      -0.983   -1.31  -1.042  ...  -1.158  -1.209  -1.208  -0.784  -0.869   \n",
       "...       ...     ...     ...  ...     ...     ...     ...     ...     ...   \n",
       "20863   0.456   0.595   0.473  ...   0.185     0.1   0.231   0.359   0.286   \n",
       "20864   0.876   1.064   0.986  ...   0.841   0.758   0.831    1.14   0.618   \n",
       "20865   0.204   0.475    0.59  ...   0.306   0.249   0.347   0.517   0.207   \n",
       "20866   0.033  -0.047  -0.357  ...  -0.108   0.028  -0.146  -0.216  -0.273   \n",
       "20867   0.201   0.336   0.348  ...  -0.169   0.102   0.124   0.096   0.255   \n",
       "\n",
       "        BXD92   BXD51   BXD48   BXD85   BXD45  \n",
       "0       0.139   0.053   0.099   0.044   0.137  \n",
       "1       -0.58  -0.187  -0.234  -0.538  -0.545  \n",
       "2       0.087   0.026   0.393   0.208   0.247  \n",
       "3       -1.91  -2.115  -1.877  -2.316  -2.343  \n",
       "5      -0.848   -1.11  -1.159  -0.857  -1.036  \n",
       "...       ...     ...     ...     ...     ...  \n",
       "20863    0.24   0.227   0.292   0.026  -0.145  \n",
       "20864   0.679   0.823   0.678    0.88   0.874  \n",
       "20865   0.322    0.33   0.246   0.301    0.23  \n",
       "20866  -0.409  -0.327  -0.215  -0.505  -0.261  \n",
       "20867    0.28   0.045    0.27   0.426   0.202  \n",
       "\n",
       "[19272 rows x 82 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove rows with no gene identifier\n",
    "liver_matrix = liver_matrix[~liver_matrix['GENE_SYMBOL'].isna()]\n",
    "liver_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Brain....."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>BXD44_F</th>\n",
       "      <th>BXD49_M</th>\n",
       "      <th>BXD44_M</th>\n",
       "      <th>BXD102_M</th>\n",
       "      <th>BXD43_M</th>\n",
       "      <th>BXD45_F</th>\n",
       "      <th>BXD45_M</th>\n",
       "      <th>BXD103_M</th>\n",
       "      <th>BXD103_F</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD92A_F</th>\n",
       "      <th>BXD70_F</th>\n",
       "      <th>BXD92A_M</th>\n",
       "      <th>BXD68_F</th>\n",
       "      <th>BXD68_M</th>\n",
       "      <th>BXD79_F</th>\n",
       "      <th>BXD85_F</th>\n",
       "      <th>BXD85_M</th>\n",
       "      <th>BXD101_M</th>\n",
       "      <th>BXD101_F</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>10338001</td>\n",
       "      <td>12.612</td>\n",
       "      <td>12.509</td>\n",
       "      <td>12.629</td>\n",
       "      <td>12.56</td>\n",
       "      <td>12.533</td>\n",
       "      <td>12.646</td>\n",
       "      <td>12.426</td>\n",
       "      <td>12.465</td>\n",
       "      <td>12.728</td>\n",
       "      <td>...</td>\n",
       "      <td>12.451</td>\n",
       "      <td>12.812</td>\n",
       "      <td>12.757</td>\n",
       "      <td>12.697</td>\n",
       "      <td>12.943</td>\n",
       "      <td>12.622</td>\n",
       "      <td>12.963</td>\n",
       "      <td>12.522</td>\n",
       "      <td>12.558</td>\n",
       "      <td>12.872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>10338003</td>\n",
       "      <td>11.187</td>\n",
       "      <td>11.043</td>\n",
       "      <td>11.116</td>\n",
       "      <td>11.104</td>\n",
       "      <td>11.09</td>\n",
       "      <td>11.151</td>\n",
       "      <td>10.955</td>\n",
       "      <td>11.06</td>\n",
       "      <td>11.256</td>\n",
       "      <td>...</td>\n",
       "      <td>10.975</td>\n",
       "      <td>11.304</td>\n",
       "      <td>11.207</td>\n",
       "      <td>11.246</td>\n",
       "      <td>11.445</td>\n",
       "      <td>11.15</td>\n",
       "      <td>11.539</td>\n",
       "      <td>10.98</td>\n",
       "      <td>11.093</td>\n",
       "      <td>11.314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>10338004</td>\n",
       "      <td>10.021</td>\n",
       "      <td>9.627</td>\n",
       "      <td>9.875</td>\n",
       "      <td>9.817</td>\n",
       "      <td>10.031</td>\n",
       "      <td>9.906</td>\n",
       "      <td>9.766</td>\n",
       "      <td>9.881</td>\n",
       "      <td>10.064</td>\n",
       "      <td>...</td>\n",
       "      <td>9.83</td>\n",
       "      <td>9.929</td>\n",
       "      <td>9.972</td>\n",
       "      <td>9.929</td>\n",
       "      <td>10.065</td>\n",
       "      <td>9.772</td>\n",
       "      <td>9.973</td>\n",
       "      <td>9.926</td>\n",
       "      <td>9.799</td>\n",
       "      <td>10.021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>10338017</td>\n",
       "      <td>13.598</td>\n",
       "      <td>13.342</td>\n",
       "      <td>13.509</td>\n",
       "      <td>13.518</td>\n",
       "      <td>13.495</td>\n",
       "      <td>13.516</td>\n",
       "      <td>13.4</td>\n",
       "      <td>13.42</td>\n",
       "      <td>13.594</td>\n",
       "      <td>...</td>\n",
       "      <td>13.562</td>\n",
       "      <td>13.723</td>\n",
       "      <td>13.667</td>\n",
       "      <td>13.543</td>\n",
       "      <td>13.811</td>\n",
       "      <td>13.482</td>\n",
       "      <td>13.651</td>\n",
       "      <td>13.559</td>\n",
       "      <td>13.575</td>\n",
       "      <td>13.827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>10338025</td>\n",
       "      <td>9.937</td>\n",
       "      <td>9.654</td>\n",
       "      <td>9.743</td>\n",
       "      <td>9.868</td>\n",
       "      <td>9.854</td>\n",
       "      <td>9.866</td>\n",
       "      <td>9.719</td>\n",
       "      <td>9.799</td>\n",
       "      <td>9.948</td>\n",
       "      <td>...</td>\n",
       "      <td>9.866</td>\n",
       "      <td>10</td>\n",
       "      <td>9.892</td>\n",
       "      <td>10.007</td>\n",
       "      <td>10.092</td>\n",
       "      <td>9.809</td>\n",
       "      <td>10.074</td>\n",
       "      <td>9.828</td>\n",
       "      <td>9.967</td>\n",
       "      <td>10.035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34789</th>\n",
       "      <td>10608720</td>\n",
       "      <td>9</td>\n",
       "      <td>9.033</td>\n",
       "      <td>9.154</td>\n",
       "      <td>8.93</td>\n",
       "      <td>9.274</td>\n",
       "      <td>8.834</td>\n",
       "      <td>9.065</td>\n",
       "      <td>9.121</td>\n",
       "      <td>9.103</td>\n",
       "      <td>...</td>\n",
       "      <td>9.085</td>\n",
       "      <td>9.136</td>\n",
       "      <td>9.199</td>\n",
       "      <td>9.265</td>\n",
       "      <td>9.126</td>\n",
       "      <td>9.092</td>\n",
       "      <td>9.342</td>\n",
       "      <td>8.994</td>\n",
       "      <td>9.025</td>\n",
       "      <td>9.337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34790</th>\n",
       "      <td>10608721</td>\n",
       "      <td>9.121</td>\n",
       "      <td>9.608</td>\n",
       "      <td>9.108</td>\n",
       "      <td>9.218</td>\n",
       "      <td>8.926</td>\n",
       "      <td>9.228</td>\n",
       "      <td>9.177</td>\n",
       "      <td>9.078</td>\n",
       "      <td>9.042</td>\n",
       "      <td>...</td>\n",
       "      <td>9.09</td>\n",
       "      <td>9.185</td>\n",
       "      <td>8.891</td>\n",
       "      <td>9.319</td>\n",
       "      <td>9.483</td>\n",
       "      <td>9.093</td>\n",
       "      <td>8.923</td>\n",
       "      <td>9.135</td>\n",
       "      <td>9.028</td>\n",
       "      <td>8.967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34791</th>\n",
       "      <td>10608722</td>\n",
       "      <td>5.691</td>\n",
       "      <td>5.466</td>\n",
       "      <td>5.591</td>\n",
       "      <td>5.674</td>\n",
       "      <td>5.831</td>\n",
       "      <td>5.541</td>\n",
       "      <td>5.618</td>\n",
       "      <td>5.604</td>\n",
       "      <td>5.553</td>\n",
       "      <td>...</td>\n",
       "      <td>5.522</td>\n",
       "      <td>5.379</td>\n",
       "      <td>5.525</td>\n",
       "      <td>5.573</td>\n",
       "      <td>5.323</td>\n",
       "      <td>5.61</td>\n",
       "      <td>5.621</td>\n",
       "      <td>5.657</td>\n",
       "      <td>5.591</td>\n",
       "      <td>5.336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34792</th>\n",
       "      <td>10608723</td>\n",
       "      <td>5.368</td>\n",
       "      <td>5.335</td>\n",
       "      <td>5.23</td>\n",
       "      <td>5.245</td>\n",
       "      <td>5.373</td>\n",
       "      <td>5.199</td>\n",
       "      <td>5.255</td>\n",
       "      <td>5.282</td>\n",
       "      <td>5.161</td>\n",
       "      <td>...</td>\n",
       "      <td>5.109</td>\n",
       "      <td>5.229</td>\n",
       "      <td>5.157</td>\n",
       "      <td>5.057</td>\n",
       "      <td>5.181</td>\n",
       "      <td>5.139</td>\n",
       "      <td>5.157</td>\n",
       "      <td>5.079</td>\n",
       "      <td>5.142</td>\n",
       "      <td>4.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34793</th>\n",
       "      <td>10608724</td>\n",
       "      <td>6.874</td>\n",
       "      <td>6.968</td>\n",
       "      <td>6.851</td>\n",
       "      <td>7.058</td>\n",
       "      <td>6.764</td>\n",
       "      <td>6.854</td>\n",
       "      <td>6.945</td>\n",
       "      <td>6.889</td>\n",
       "      <td>6.752</td>\n",
       "      <td>...</td>\n",
       "      <td>6.853</td>\n",
       "      <td>6.921</td>\n",
       "      <td>6.866</td>\n",
       "      <td>6.789</td>\n",
       "      <td>6.996</td>\n",
       "      <td>6.973</td>\n",
       "      <td>6.994</td>\n",
       "      <td>6.801</td>\n",
       "      <td>6.777</td>\n",
       "      <td>6.966</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34760 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ID BXD44_F BXD49_M BXD44_M BXD102_M BXD43_M BXD45_F BXD45_M  \\\n",
       "34     10338001  12.612  12.509  12.629    12.56  12.533  12.646  12.426   \n",
       "35     10338003  11.187  11.043  11.116   11.104   11.09  11.151  10.955   \n",
       "36     10338004  10.021   9.627   9.875    9.817  10.031   9.906   9.766   \n",
       "37     10338017  13.598  13.342  13.509   13.518  13.495  13.516    13.4   \n",
       "38     10338025   9.937   9.654   9.743    9.868   9.854   9.866   9.719   \n",
       "...         ...     ...     ...     ...      ...     ...     ...     ...   \n",
       "34789  10608720       9   9.033   9.154     8.93   9.274   8.834   9.065   \n",
       "34790  10608721   9.121   9.608   9.108    9.218   8.926   9.228   9.177   \n",
       "34791  10608722   5.691   5.466   5.591    5.674   5.831   5.541   5.618   \n",
       "34792  10608723   5.368   5.335    5.23    5.245   5.373   5.199   5.255   \n",
       "34793  10608724   6.874   6.968   6.851    7.058   6.764   6.854   6.945   \n",
       "\n",
       "      BXD103_M BXD103_F  ... BXD92A_F BXD70_F BXD92A_M BXD68_F BXD68_M  \\\n",
       "34      12.465   12.728  ...   12.451  12.812   12.757  12.697  12.943   \n",
       "35       11.06   11.256  ...   10.975  11.304   11.207  11.246  11.445   \n",
       "36       9.881   10.064  ...     9.83   9.929    9.972   9.929  10.065   \n",
       "37       13.42   13.594  ...   13.562  13.723   13.667  13.543  13.811   \n",
       "38       9.799    9.948  ...    9.866      10    9.892  10.007  10.092   \n",
       "...        ...      ...  ...      ...     ...      ...     ...     ...   \n",
       "34789    9.121    9.103  ...    9.085   9.136    9.199   9.265   9.126   \n",
       "34790    9.078    9.042  ...     9.09   9.185    8.891   9.319   9.483   \n",
       "34791    5.604    5.553  ...    5.522   5.379    5.525   5.573   5.323   \n",
       "34792    5.282    5.161  ...    5.109   5.229    5.157   5.057   5.181   \n",
       "34793    6.889    6.752  ...    6.853   6.921    6.866   6.789   6.996   \n",
       "\n",
       "      BXD79_F BXD85_F BXD85_M BXD101_M BXD101_F  \n",
       "34     12.622  12.963  12.522   12.558   12.872  \n",
       "35      11.15  11.539   10.98   11.093   11.314  \n",
       "36      9.772   9.973   9.926    9.799   10.021  \n",
       "37     13.482  13.651  13.559   13.575   13.827  \n",
       "38      9.809  10.074   9.828    9.967   10.035  \n",
       "...       ...     ...     ...      ...      ...  \n",
       "34789   9.092   9.342   8.994    9.025    9.337  \n",
       "34790   9.093   8.923   9.135    9.028    8.967  \n",
       "34791    5.61   5.621   5.657    5.591    5.336  \n",
       "34792   5.139   5.157   5.079    5.142     4.94  \n",
       "34793   6.973   6.994   6.801    6.777    6.966  \n",
       "\n",
       "[34760 rows x 90 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove metadata rows to get the raw data only\n",
    "brain_exp = brain_exp[~brain_exp['!Sample_title'].str.contains('!Sample_', na=False)]\n",
    "brain_exp = brain_exp[~brain_exp['!Sample_title'].str.contains('!series_', na=False)]\n",
    "brain_exp = brain_exp[brain_exp['!Sample_title'] != 'ID_REF']\n",
    "\n",
    "# Rename ID column to match the annotation matrix\n",
    "brain_exp = brain_exp.rename(columns = {'!Sample_title' : 'ID'})\n",
    "brain_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
