{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Systems genetics 2020 - Final Project  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Note!\n",
    "Our project is devided into several python modules which are loaded and excecuted from this notebook. <br>\n",
    "Most of our code's project can be found under the /modules directory. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import public packges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import f\n",
    "import matplotlib.pyplot as plt\n",
    "import os.path\n",
    "import sys\n",
    "from scipy import stats \n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import custom packges (our python modules)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('modules')\n",
    "\n",
    "from utils import print_dim, print_stats\n",
    "from preprocessing import data_annotations_merge\n",
    "from regression import eqtl_analysis, qtl_analysis\n",
    "from results_analysis import get_associations, cis_trans_annotation\n",
    "from causality_analysis import generate_models_df, get_LR, get_shuffled_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "bold_s = '\\033[1m'\n",
    "bold_e = '\\033[0m'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "liver_exp = pd.read_csv('data/GSE17522_series_matrix_liver.txt', sep = \"\\t\")\n",
    "brain_exp = pd.read_csv('data/GSE36674_series_matrix_hypothalamus.txt', sep = \"\\t\")\n",
    "\n",
    "liver_annotations = pd.read_csv('data/annotations_liver_GPL6466-9752.txt', sep = \"\\t\")\n",
    "brain_annotations = pd.read_csv('data/annotation_brain.annot', sep = \"\\t\")\n",
    "\n",
    "genotypes = pd.read_excel('data/genotypes.xls', headers=None)\n",
    "phenotypes = pd.read_excel('data/phenotypes.xls')\n",
    "\n",
    "mgi = pd.read_csv('data/MGI_Coordinates.Build37.rpt.txt', sep = \"\\t\", error_bad_lines=False, warn_bad_lines=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Gene expression data preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### A. Merge data file with annotation file to get your input matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "brain_annotations = brain_annotations.rename(columns={'Gene symbol' : 'GENE_SYMBOL'})\n",
    "brain_matrix = data_annotations_merge(brain_exp, brain_annotations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Liver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "liver_matrix = data_annotations_merge(liver_exp, liver_annotations)\n",
    "\n",
    "# Keep BXD columns and identifier columns only\n",
    "liver_bxd_cols = list(liver_matrix.filter(regex=(\"BXD*\")).columns)\n",
    "id_cols = ['ID', 'GENE_SYMBOL']\n",
    "liver_matrix = liver_matrix[id_cols + liver_bxd_cols]\n",
    "\n",
    "# Rename BXD columns\n",
    "liver_matrix = liver_matrix.rename(columns={col : col.split('_')[1] + '_' + col.split('_')[2] \\\n",
    "                                            for col in liver_matrix.columns.drop(id_cols)})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### B. Keep only BXD columns that exists in the four files: genotypes, phenotypes and each tissue seperately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_bxd_cols = list(genotypes.filter(regex=(\"BXD*\")).columns)\n",
    "phenotypes_bxd_cols = list(phenotypes.filter(regex=(\"BXD*\")).columns)\n",
    "\n",
    "genotypes_baseline_cols = [\"Locus\", \"Chr_Build37\", \"Build37_position\"]\n",
    "phenotypes_baseline_cols = [\"Phenotype\", \"Authors\", \"Year\", \"Pubmed Id\"]\n",
    "\n",
    "# liver common columns\n",
    "liver_matrix = liver_matrix.rename(columns = {col : col.split('_')[0] for col in liver_matrix.drop(columns = id_cols)}) \n",
    "liver_bxd_cols = list(liver_matrix.filter(regex='BXD').columns)\n",
    "liver_common_bxd = list(set(genotypes_bxd_cols) & set(phenotypes_bxd_cols) &  set(liver_bxd_cols))\n",
    "liver_common_bxd.sort()\n",
    "liver_matrix = liver_matrix[['GENE_SYMBOL'] + liver_common_bxd]\n",
    "liver_genotypes = genotypes[genotypes_baseline_cols + liver_common_bxd]\n",
    "liver_phenotypes = phenotypes[phenotypes_baseline_cols + liver_common_bxd]\n",
    "\n",
    "# brain common columns\n",
    "brain_matrix = brain_matrix.rename(columns = {col : col.split('_')[0] for col in brain_matrix.drop(columns = id_cols)}) \n",
    "brain_bxd_cols = list(brain_matrix.filter(regex='BXD').columns)\n",
    "brain_common_bxd = list(set(genotypes_bxd_cols) & set(phenotypes_bxd_cols) &  set(brain_bxd_cols))\n",
    "brain_common_bxd.sort()\n",
    "brain_matrix = brain_matrix[['GENE_SYMBOL'] + brain_common_bxd]\n",
    "brain_genotypes = genotypes[genotypes_baseline_cols + brain_common_bxd]\n",
    "brain_phenotypes = phenotypes[phenotypes_baseline_cols + brain_common_bxd]\n",
    "\n",
    "# drop empty lines from phenotypes file\n",
    "brain_phenotypes = brain_phenotypes.iloc[brain_phenotypes[brain_common_bxd].dropna(how='all').index]\n",
    "liver_phenotypes = liver_phenotypes.iloc[liver_phenotypes[liver_common_bxd].dropna(how='all').index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### C. Use only representative genomic loci - Drop duplicated rows (of neighboring loci)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LIVER Genotypes: Drop duplications:\n",
      "#SNPs before: 3796   -->   #SNPs after: 1403\n",
      "LIVER Genotypes: Drop duplications:\n",
      "#SNPs before: 3796   -->   #SNPs after: 1598\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Locus</th>\n",
       "      <th>Chr_Build37</th>\n",
       "      <th>Build37_position</th>\n",
       "      <th>BXD1</th>\n",
       "      <th>BXD100</th>\n",
       "      <th>BXD101</th>\n",
       "      <th>BXD102</th>\n",
       "      <th>BXD103</th>\n",
       "      <th>BXD11</th>\n",
       "      <th>BXD12</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD80</th>\n",
       "      <th>BXD83</th>\n",
       "      <th>BXD84</th>\n",
       "      <th>BXD85</th>\n",
       "      <th>BXD87</th>\n",
       "      <th>BXD89</th>\n",
       "      <th>BXD90</th>\n",
       "      <th>BXD95</th>\n",
       "      <th>BXD97</th>\n",
       "      <th>BXD99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs6269442</td>\n",
       "      <td>1</td>\n",
       "      <td>3482276</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>H</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs6376963</td>\n",
       "      <td>1</td>\n",
       "      <td>5008090</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>H</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs3677817</td>\n",
       "      <td>1</td>\n",
       "      <td>5176059</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>H</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs8236463</td>\n",
       "      <td>1</td>\n",
       "      <td>5579194</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>H</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs6298633</td>\n",
       "      <td>1</td>\n",
       "      <td>6820242</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>H</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Locus  Chr_Build37  Build37_position BXD1 BXD100 BXD101 BXD102 BXD103  \\\n",
       "0  rs6269442            1           3482276    B      B      U      U      U   \n",
       "2  rs6376963            1           5008090    B      B      U      U      U   \n",
       "3  rs3677817            1           5176059    B      B      U      U      U   \n",
       "4  rs8236463            1           5579194    B      B      U      U      U   \n",
       "6  rs6298633            1           6820242    B      B      U      U      U   \n",
       "\n",
       "  BXD11 BXD12  ... BXD80 BXD83 BXD84 BXD85 BXD87 BXD89 BXD90 BXD95 BXD97 BXD99  \n",
       "0     B     D  ...     D     H     B     D     B     B     B     D     B     B  \n",
       "2     B     D  ...     D     H     B     D     B     B     B     D     B     B  \n",
       "3     B     D  ...     D     H     B     D     B     B     B     D     B     B  \n",
       "4     D     D  ...     D     H     B     D     B     B     B     D     B     B  \n",
       "6     D     D  ...     D     H     B     D     B     B     B     D     B     B  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# liver genotypes filtering\n",
    "snps_org = len(liver_genotypes)\n",
    "\n",
    "bxd_data = liver_genotypes[liver_common_bxd]\n",
    "duplicates_indx = bxd_data[bxd_data.shift() != bxd_data].dropna(how='all').index\n",
    "liver_genotypes = liver_genotypes.loc[duplicates_indx]\n",
    "\n",
    "print(\"LIVER Genotypes: Drop duplications:\\n#SNPs before: %d   -->   #SNPs after: %d\" % (snps_org, len(liver_genotypes)))\n",
    "liver_genotypes.head(5)\n",
    "\n",
    "# Brain genotypes filtering\n",
    "snps_org = len(brain_genotypes)\n",
    "\n",
    "bxd_data = brain_genotypes[brain_common_bxd]\n",
    "duplicates_indx = bxd_data[bxd_data.shift() != bxd_data].dropna(how='all').index\n",
    "brain_genotypes = brain_genotypes.loc[duplicates_indx]\n",
    "\n",
    "print(\"LIVER Genotypes: Drop duplications:\\n#SNPs before: %d   -->   #SNPs after: %d\" % (snps_org, len(brain_genotypes)))\n",
    "brain_genotypes.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### • Remove rows with no gene identifier, <br> • Remove rows with low maximal value.  <br> • Remove rows with low variance.  <br> • Average multiple rows\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Liver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num of rows after removing all rows with maximal values less than 0.70: 7772 \n",
      "Num of rows after removing all rows with variance less than 0.20: 1308 \n",
      "Num of rows after removing duplicated rows: %d  1285\n"
     ]
    }
   ],
   "source": [
    "max_treshold = 0.7\n",
    "var_treshold = 0.2\n",
    "\n",
    "# Remove rows with no gene identifier\n",
    "liver_matrix = liver_matrix[~liver_matrix['GENE_SYMBOL'].isna()]\n",
    "liver_bxd_cols = list(liver_matrix.filter(regex=(\"BXD*\")).columns)\n",
    "\n",
    "for col in liver_bxd_cols:\n",
    "    liver_matrix[col] = liver_matrix[col].astype('float64')\n",
    "\n",
    "# Filter by maximal value\n",
    "liver_matrix['max'] = liver_matrix.drop(columns=['GENE_SYMBOL']).max(axis=1)\n",
    "liver_matrix = liver_matrix[liver_matrix['max'] >= max_treshold]\n",
    "liver_matrix = liver_matrix.drop(columns = 'max')\n",
    "print(\"Num of rows after removing all rows with maximal values less than %.2f: %d \" % (max_treshold, len(liver_matrix)))\n",
    "\n",
    "# Filter by variance\n",
    "liver_matrix['var'] = liver_matrix.drop(columns=['GENE_SYMBOL']).var(axis=1)\n",
    "liver_matrix = liver_matrix[liver_matrix['var'] >= var_treshold]\n",
    "liver_matrix = liver_matrix.drop(columns = 'var')\n",
    "print(\"Num of rows after removing all rows with variance less than %.2f: %d \" % (var_treshold, len(liver_matrix)))\n",
    "\n",
    "# Group multiple rows by mean\n",
    "liver_matrix = liver_matrix.groupby('GENE_SYMBOL').agg('mean').reset_index()\n",
    "print(\"Num of rows after removing duplicated rows: %d \", len(liver_matrix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num of rows after removing all rows with maximal values less than 9.00: 9893 \n",
      "Num of rows after removing all rows with variance less than 0.04: 1444 \n",
      "Num of rows after removing duplicated rows: %d  1247\n"
     ]
    }
   ],
   "source": [
    "max_treshold = 9\n",
    "var_treshold = 0.04\n",
    "\n",
    "# Remove rows with no gene identifier\n",
    "brain_matrix = brain_matrix[~brain_matrix['GENE_SYMBOL'].isna()]\n",
    "brain_bxd_cols = list(brain_matrix.filter(regex=(\"BXD*\")).columns)\n",
    "\n",
    "for col in brain_bxd_cols:\n",
    "    brain_matrix[col] = brain_matrix[col].astype('float64')\n",
    "\n",
    "# Filter by maximal value\n",
    "brain_matrix['max'] = brain_matrix.drop(columns=['GENE_SYMBOL']).max(axis=1)\n",
    "brain_matrix = brain_matrix[brain_matrix['max'] >= max_treshold]\n",
    "brain_matrix = brain_matrix.drop(columns = 'max')\n",
    "print(\"Num of rows after removing all rows with maximal values less than %.2f: %d \" % (max_treshold, len(brain_matrix)))\n",
    "\n",
    "# Filter by variance\n",
    "brain_matrix['var'] = brain_matrix.drop(columns=['GENE_SYMBOL']).var(axis=1)\n",
    "brain_matrix = brain_matrix[brain_matrix['var'] >= var_treshold]\n",
    "brain_matrix = brain_matrix.drop(columns = 'var')\n",
    "print(\"Num of rows after removing all rows with variance less than %.2f: %d \" % (var_treshold, len(brain_matrix)))\n",
    "\n",
    "# Group multiple rows by mean\n",
    "brain_matrix = brain_matrix.groupby('GENE_SYMBOL').agg('mean').reset_index()\n",
    "print(\"Num of rows after removing duplicated rows: %d \", len(brain_matrix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### • Average across different individuals of the same strain (Females and males)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "liver_matrix = liver_matrix.rename(columns = {col : col.split('_')[0] for col in liver_matrix.drop(columns = ['GENE_SYMBOL'])}) \n",
    "liver_matrix = liver_matrix.set_index(['GENE_SYMBOL']) \n",
    "liver_matrix = liver_matrix.groupby(by=liver_matrix.columns, axis=1).mean()\n",
    "liver_matrix = liver_matrix.reset_index()\n",
    "\n",
    "brain_matrix = brain_matrix.rename(columns = {col : col.split('_')[0] for col in brain_matrix.drop(columns = ['GENE_SYMBOL'])}) \n",
    "brain_matrix = brain_matrix.set_index(['GENE_SYMBOL']) \n",
    "brain_matrix = brain_matrix.groupby(by=brain_matrix.columns, axis=1).mean()\n",
    "brain_matrix = brain_matrix.reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. eQTL analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run  Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Liver: \n",
      "* Expression matrix size: 1285\n",
      "* Genotype matrix size: 1403\n",
      "* Expected num of tests: 1,802,855\n",
      "\n",
      "Brain: \n",
      "* Expression matrix size: 1247\n",
      "* Genotype matrix size: 1598\n",
      "* Expected num of tests: 1,992,706\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print_dim(len(liver_matrix), len(liver_genotypes), title=\"Liver: \")\n",
    "print_dim(len(brain_matrix), len(brain_genotypes), title=\"Brain: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Brain eqtl\n",
    "if not os.path.exists(\"brain_reg_results.csv\"):\n",
    "    brain_eqtls = eqtl_analysis(brain_matrix, brain_genotypes, file_prefix=\"brain_\")\n",
    "else:\n",
    "    brain_eqtls = pd.read_csv(\"brain_reg_results.csv\")\n",
    "    brain_eqtls = brain_eqtls.drop(columns='Unnamed: 0')\n",
    "    \n",
    "# Liver eqtl\n",
    "if not os.path.exists(\"liver_reg_results.csv\"):\n",
    "    liver_eqtls = eqtl_analysis(liver_matrix, liver_genotypes, file_prefix=\"liver_\")\n",
    "else:\n",
    "    liver_eqtls = pd.read_csv(\"liver_reg_results.csv\")\n",
    "    liver_eqtls = liver_eqtls.drop(columns='Unnamed: 0')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cis/Trans annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████| 1247/1247 [04:04<00:00,  5.11it/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████| 1285/1285 [04:41<00:00,  4.57it/s]\n"
     ]
    }
   ],
   "source": [
    "brain_eqtls = cis_trans_annotation(brain_eqtls.copy(), mgi)\n",
    "liver_eqtls = cis_trans_annotation(liver_eqtls.copy(), mgi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Multiple test correction and associations filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m1. Brain\u001b[0m\n",
      "\u001b[1mNumber of tests: \u001b[0m 1992706\n",
      "\n",
      "\u001b[1mNumber of different significant eQTLs: \u001b[0m 472\n",
      "From which: \n",
      " 260 - cis-acting \n",
      " 231 - trans-acting \n",
      "\n",
      "\u001b[1mNumber of total significant eQTLs: \u001b[0m 868\n",
      "From which: \n",
      " 349 - cis-acting \n",
      " 337 - trans-acting \n",
      "\u001b[1m\n",
      "\n",
      "2. Liver\u001b[0m\n",
      "\u001b[1mNumber of tests: \u001b[0m 1802855\n",
      "\n",
      "\u001b[1mNumber of different significant eQTLs: \u001b[0m 481\n",
      "From which: \n",
      " 261 - cis-acting \n",
      " 278 - trans-acting \n",
      "\n",
      "\u001b[1mNumber of total significant eQTLs: \u001b[0m 840\n",
      "From which: \n",
      " 325 - cis-acting \n",
      " 468 - trans-acting \n"
     ]
    }
   ],
   "source": [
    "brain_associations, brain_num_tests = get_associations(brain_eqtls) # by bonfferoni corrections\n",
    "brain_associations.to_csv(\"brain_assoc_eqtl.csv\")\n",
    "\n",
    "liver_associations, liver_num_tests = get_associations(liver_eqtls) # by bonfferoni corrections\n",
    "liver_associations.to_csv(\"liver_assoc_eqtl.csv\")\n",
    "\n",
    "print_stats(brain_associations, brain_num_tests, \"1. Brain\")\n",
    "print_stats(liver_associations, liver_num_tests, \"\\n\\n2. Liver\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. QTL analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Currently looks like no morphine phenotype is significant.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>gene</th>\n",
       "      <th>p-value</th>\n",
       "      <th>minus_log_p-value</th>\n",
       "      <th>closeness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10625</th>\n",
       "      <td>rs13481466</td>\n",
       "      <td>12</td>\n",
       "      <td>57107315</td>\n",
       "      <td>1700047I17Rik2///Fam177a</td>\n",
       "      <td>1.614445e-08</td>\n",
       "      <td>7.791977</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21243</th>\n",
       "      <td>rs6163111</td>\n",
       "      <td>5</td>\n",
       "      <td>74219748</td>\n",
       "      <td>2310040G07Rik</td>\n",
       "      <td>2.270665e-10</td>\n",
       "      <td>9.643847</td>\n",
       "      <td>trans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21244</th>\n",
       "      <td>rs3153753</td>\n",
       "      <td>5</td>\n",
       "      <td>76574922</td>\n",
       "      <td>2310040G07Rik</td>\n",
       "      <td>1.110223e-16</td>\n",
       "      <td>15.954590</td>\n",
       "      <td>cis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21245</th>\n",
       "      <td>CEL-5_76522694</td>\n",
       "      <td>5</td>\n",
       "      <td>78208834</td>\n",
       "      <td>2310040G07Rik</td>\n",
       "      <td>3.034129e-12</td>\n",
       "      <td>11.517966</td>\n",
       "      <td>cis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21246</th>\n",
       "      <td>rs13478357</td>\n",
       "      <td>5</td>\n",
       "      <td>82835236</td>\n",
       "      <td>2310040G07Rik</td>\n",
       "      <td>1.342021e-10</td>\n",
       "      <td>9.872241</td>\n",
       "      <td>trans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979677</th>\n",
       "      <td>rs13482864</td>\n",
       "      <td>17</td>\n",
       "      <td>9565936</td>\n",
       "      <td>Zfp983</td>\n",
       "      <td>8.058357e-10</td>\n",
       "      <td>9.093753</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979678</th>\n",
       "      <td>rs13482868</td>\n",
       "      <td>17</td>\n",
       "      <td>10586984</td>\n",
       "      <td>Zfp983</td>\n",
       "      <td>7.858147e-11</td>\n",
       "      <td>10.104680</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979679</th>\n",
       "      <td>rs13482873</td>\n",
       "      <td>17</td>\n",
       "      <td>11606730</td>\n",
       "      <td>Zfp983</td>\n",
       "      <td>1.028067e-13</td>\n",
       "      <td>12.987979</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979680</th>\n",
       "      <td>CEL-17_20354682</td>\n",
       "      <td>17</td>\n",
       "      <td>21827381</td>\n",
       "      <td>Zfp983</td>\n",
       "      <td>4.858505e-10</td>\n",
       "      <td>9.313497</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1988610</th>\n",
       "      <td>D7Mit371</td>\n",
       "      <td>7</td>\n",
       "      <td>139448713</td>\n",
       "      <td>Zranb1</td>\n",
       "      <td>1.637359e-08</td>\n",
       "      <td>7.785856</td>\n",
       "      <td>cis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>868 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     SNP  chromosome   position                      gene  \\\n",
       "10625         rs13481466          12   57107315  1700047I17Rik2///Fam177a   \n",
       "21243          rs6163111           5   74219748             2310040G07Rik   \n",
       "21244          rs3153753           5   76574922             2310040G07Rik   \n",
       "21245     CEL-5_76522694           5   78208834             2310040G07Rik   \n",
       "21246         rs13478357           5   82835236             2310040G07Rik   \n",
       "...                  ...         ...        ...                       ...   \n",
       "1979677       rs13482864          17    9565936                    Zfp983   \n",
       "1979678       rs13482868          17   10586984                    Zfp983   \n",
       "1979679       rs13482873          17   11606730                    Zfp983   \n",
       "1979680  CEL-17_20354682          17   21827381                    Zfp983   \n",
       "1988610         D7Mit371           7  139448713                    Zranb1   \n",
       "\n",
       "              p-value  minus_log_p-value closeness  \n",
       "10625    1.614445e-08           7.791977   Unknown  \n",
       "21243    2.270665e-10           9.643847     trans  \n",
       "21244    1.110223e-16          15.954590       cis  \n",
       "21245    3.034129e-12          11.517966       cis  \n",
       "21246    1.342021e-10           9.872241     trans  \n",
       "...               ...                ...       ...  \n",
       "1979677  8.058357e-10           9.093753   Unknown  \n",
       "1979678  7.858147e-11          10.104680   Unknown  \n",
       "1979679  1.028067e-13          12.987979   Unknown  \n",
       "1979680  4.858505e-10           9.313497   Unknown  \n",
       "1988610  1.637359e-08           7.785856       cis  \n",
       "\n",
       "[868 rows x 7 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#brain_phenotypes.to_excel('lool_brain.xlsx`\n",
    "# brain_associations.join(brain_genotypes, on = 'Locus')\n",
    "# brain_associations.head()\n",
    "temp = pd.merge(brain_associations['SNP'], brain_genotypes, how='left', left_on=['SNP'], right_on=['Locus']).drop('SNP', axis=1)\n",
    "temp[temp['Locus'].duplicated()==True]\n",
    "\n",
    "brain_associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Year</th>\n",
       "      <th>Pubmed Id</th>\n",
       "      <th>BXD1</th>\n",
       "      <th>BXD11</th>\n",
       "      <th>BXD12</th>\n",
       "      <th>BXD13</th>\n",
       "      <th>BXD14</th>\n",
       "      <th>BXD15</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD60</th>\n",
       "      <th>BXD62</th>\n",
       "      <th>BXD69</th>\n",
       "      <th>BXD73</th>\n",
       "      <th>BXD77</th>\n",
       "      <th>BXD8</th>\n",
       "      <th>BXD85</th>\n",
       "      <th>BXD86</th>\n",
       "      <th>BXD9</th>\n",
       "      <th>BXD92</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Morphine response (dose ip), hypothermia [slop...</td>\n",
       "      <td>Belknap JK, Crabbe JC</td>\n",
       "      <td>1992</td>\n",
       "      <td>1632590.0</td>\n",
       "      <td>-7.060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.020</td>\n",
       "      <td>-3.430</td>\n",
       "      <td>-3.750</td>\n",
       "      <td>-1.8800</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-7.36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.930</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Morphine response (dose ip), analgesia, slope ...</td>\n",
       "      <td>Belknap JK, Crabbe JC</td>\n",
       "      <td>1992</td>\n",
       "      <td>1632590.0</td>\n",
       "      <td>50.100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33.900</td>\n",
       "      <td>25.800</td>\n",
       "      <td>31.200</td>\n",
       "      <td>25.5000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62.70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27.600</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Morphine response (dose mg/kg ip), Straub tail...</td>\n",
       "      <td>Belknap JK, Crabbe JC</td>\n",
       "      <td>1992</td>\n",
       "      <td>1632590.0</td>\n",
       "      <td>1.370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.810</td>\n",
       "      <td>0.940</td>\n",
       "      <td>1.370</td>\n",
       "      <td>0.9100</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Morphine response (16 mg/kg), analgesia relati...</td>\n",
       "      <td>Belknap JK, Mogil JS, Helms ML, Richards SP, O...</td>\n",
       "      <td>1995</td>\n",
       "      <td>7643715.0</td>\n",
       "      <td>63.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.000</td>\n",
       "      <td>20.000</td>\n",
       "      <td>23.000</td>\n",
       "      <td>32.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>Morphine response (0.3 to 0.7 mg/ml p.o.), con...</td>\n",
       "      <td>Phillips TJ, Belknap JK, Crabbe JC</td>\n",
       "      <td>1991</td>\n",
       "      <td>2065120.0</td>\n",
       "      <td>16.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.000</td>\n",
       "      <td>29.000</td>\n",
       "      <td>23.000</td>\n",
       "      <td>148.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1517</th>\n",
       "      <td>Morphine response (50 mg/kg ip), vertical acti...</td>\n",
       "      <td>Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...</td>\n",
       "      <td>2010</td>\n",
       "      <td>19958391.0</td>\n",
       "      <td>469.538</td>\n",
       "      <td>227.143</td>\n",
       "      <td>382.571</td>\n",
       "      <td>103.222</td>\n",
       "      <td>322.500</td>\n",
       "      <td>61.2500</td>\n",
       "      <td>...</td>\n",
       "      <td>2.07692</td>\n",
       "      <td>68.6667</td>\n",
       "      <td>42.5000</td>\n",
       "      <td>142.000</td>\n",
       "      <td>79.75</td>\n",
       "      <td>9.00</td>\n",
       "      <td>435.250</td>\n",
       "      <td>228.8</td>\n",
       "      <td>256.909</td>\n",
       "      <td>306.333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518</th>\n",
       "      <td>Morphine response (50 mg/kg ip), vertical acti...</td>\n",
       "      <td>Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...</td>\n",
       "      <td>2010</td>\n",
       "      <td>19958391.0</td>\n",
       "      <td>397.154</td>\n",
       "      <td>321.143</td>\n",
       "      <td>603.000</td>\n",
       "      <td>225.778</td>\n",
       "      <td>450.375</td>\n",
       "      <td>53.0833</td>\n",
       "      <td>...</td>\n",
       "      <td>5.53846</td>\n",
       "      <td>147.3330</td>\n",
       "      <td>54.5714</td>\n",
       "      <td>206.188</td>\n",
       "      <td>33.00</td>\n",
       "      <td>67.50</td>\n",
       "      <td>516.167</td>\n",
       "      <td>338.4</td>\n",
       "      <td>422.727</td>\n",
       "      <td>361.444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1519</th>\n",
       "      <td>Morphine response (50 mg/kg ip), vertical acti...</td>\n",
       "      <td>Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...</td>\n",
       "      <td>2010</td>\n",
       "      <td>19958391.0</td>\n",
       "      <td>512.154</td>\n",
       "      <td>349.000</td>\n",
       "      <td>656.286</td>\n",
       "      <td>280.556</td>\n",
       "      <td>581.625</td>\n",
       "      <td>49.1667</td>\n",
       "      <td>...</td>\n",
       "      <td>14.53850</td>\n",
       "      <td>188.7500</td>\n",
       "      <td>83.0000</td>\n",
       "      <td>323.875</td>\n",
       "      <td>16.75</td>\n",
       "      <td>99.50</td>\n",
       "      <td>618.417</td>\n",
       "      <td>503.3</td>\n",
       "      <td>583.455</td>\n",
       "      <td>373.111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>Morphine response (50 mg/kg ip), vertical acti...</td>\n",
       "      <td>Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...</td>\n",
       "      <td>2010</td>\n",
       "      <td>19958391.0</td>\n",
       "      <td>584.923</td>\n",
       "      <td>332.571</td>\n",
       "      <td>603.286</td>\n",
       "      <td>320.778</td>\n",
       "      <td>790.625</td>\n",
       "      <td>47.0833</td>\n",
       "      <td>...</td>\n",
       "      <td>20.46150</td>\n",
       "      <td>192.9170</td>\n",
       "      <td>113.7140</td>\n",
       "      <td>391.875</td>\n",
       "      <td>47.00</td>\n",
       "      <td>192.50</td>\n",
       "      <td>627.333</td>\n",
       "      <td>641.8</td>\n",
       "      <td>651.909</td>\n",
       "      <td>375.111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1521</th>\n",
       "      <td>Morphine response (50 mg/kg ip), vertical acti...</td>\n",
       "      <td>Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...</td>\n",
       "      <td>2010</td>\n",
       "      <td>19958391.0</td>\n",
       "      <td>612.000</td>\n",
       "      <td>286.857</td>\n",
       "      <td>575.143</td>\n",
       "      <td>307.444</td>\n",
       "      <td>828.875</td>\n",
       "      <td>56.5833</td>\n",
       "      <td>...</td>\n",
       "      <td>2.00000</td>\n",
       "      <td>203.9170</td>\n",
       "      <td>129.0000</td>\n",
       "      <td>466.625</td>\n",
       "      <td>68.00</td>\n",
       "      <td>228.25</td>\n",
       "      <td>572.500</td>\n",
       "      <td>697.1</td>\n",
       "      <td>635.909</td>\n",
       "      <td>360.667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Phenotype  \\\n",
       "15    Morphine response (dose ip), hypothermia [slop...   \n",
       "16    Morphine response (dose ip), analgesia, slope ...   \n",
       "17    Morphine response (dose mg/kg ip), Straub tail...   \n",
       "28    Morphine response (16 mg/kg), analgesia relati...   \n",
       "466   Morphine response (0.3 to 0.7 mg/ml p.o.), con...   \n",
       "...                                                 ...   \n",
       "1517  Morphine response (50 mg/kg ip), vertical acti...   \n",
       "1518  Morphine response (50 mg/kg ip), vertical acti...   \n",
       "1519  Morphine response (50 mg/kg ip), vertical acti...   \n",
       "1520  Morphine response (50 mg/kg ip), vertical acti...   \n",
       "1521  Morphine response (50 mg/kg ip), vertical acti...   \n",
       "\n",
       "                                                Authors  Year   Pubmed Id  \\\n",
       "15                                Belknap JK, Crabbe JC  1992   1632590.0   \n",
       "16                                Belknap JK, Crabbe JC  1992   1632590.0   \n",
       "17                                Belknap JK, Crabbe JC  1992   1632590.0   \n",
       "28    Belknap JK, Mogil JS, Helms ML, Richards SP, O...  1995   7643715.0   \n",
       "466                  Phillips TJ, Belknap JK, Crabbe JC  1991   2065120.0   \n",
       "...                                                 ...   ...         ...   \n",
       "1517  Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...  2010  19958391.0   \n",
       "1518  Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...  2010  19958391.0   \n",
       "1519  Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...  2010  19958391.0   \n",
       "1520  Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...  2010  19958391.0   \n",
       "1521  Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre ...  2010  19958391.0   \n",
       "\n",
       "         BXD1    BXD11    BXD12    BXD13    BXD14     BXD15  ...     BXD60  \\\n",
       "15     -7.060      NaN   -2.020   -3.430   -3.750   -1.8800  ...       NaN   \n",
       "16     50.100      NaN   33.900   25.800   31.200   25.5000  ...       NaN   \n",
       "17      1.370      NaN    0.810    0.940    1.370    0.9100  ...       NaN   \n",
       "28     63.000      NaN   10.000   20.000   23.000   32.0000  ...       NaN   \n",
       "466    16.000      NaN   97.000   29.000   23.000  148.0000  ...       NaN   \n",
       "...       ...      ...      ...      ...      ...       ...  ...       ...   \n",
       "1517  469.538  227.143  382.571  103.222  322.500   61.2500  ...   2.07692   \n",
       "1518  397.154  321.143  603.000  225.778  450.375   53.0833  ...   5.53846   \n",
       "1519  512.154  349.000  656.286  280.556  581.625   49.1667  ...  14.53850   \n",
       "1520  584.923  332.571  603.286  320.778  790.625   47.0833  ...  20.46150   \n",
       "1521  612.000  286.857  575.143  307.444  828.875   56.5833  ...   2.00000   \n",
       "\n",
       "         BXD62     BXD69    BXD73  BXD77    BXD8    BXD85  BXD86     BXD9  \\\n",
       "15         NaN       NaN      NaN    NaN   -7.36      NaN    NaN   -2.930   \n",
       "16         NaN       NaN      NaN    NaN   62.70      NaN    NaN   27.600   \n",
       "17         NaN       NaN      NaN    NaN    0.42      NaN    NaN    0.750   \n",
       "28         NaN       NaN      NaN    NaN   72.00      NaN    NaN   50.000   \n",
       "466        NaN       NaN      NaN    NaN   27.00      NaN    NaN   27.000   \n",
       "...        ...       ...      ...    ...     ...      ...    ...      ...   \n",
       "1517   68.6667   42.5000  142.000  79.75    9.00  435.250  228.8  256.909   \n",
       "1518  147.3330   54.5714  206.188  33.00   67.50  516.167  338.4  422.727   \n",
       "1519  188.7500   83.0000  323.875  16.75   99.50  618.417  503.3  583.455   \n",
       "1520  192.9170  113.7140  391.875  47.00  192.50  627.333  641.8  651.909   \n",
       "1521  203.9170  129.0000  466.625  68.00  228.25  572.500  697.1  635.909   \n",
       "\n",
       "        BXD92  \n",
       "15        NaN  \n",
       "16        NaN  \n",
       "17        NaN  \n",
       "28        NaN  \n",
       "466       NaN  \n",
       "...       ...  \n",
       "1517  306.333  \n",
       "1518  361.444  \n",
       "1519  373.111  \n",
       "1520  375.111  \n",
       "1521  360.667  \n",
       "\n",
       "[115 rows x 44 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "liver_phenotypes[liver_phenotypes['Phenotype'].str.contains(\"Morphine\")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████| 115/115 [09:57<00:00,  5.20s/it]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████| 115/115 [07:31<00:00,  3.92s/it]\n"
     ]
    }
   ],
   "source": [
    "logpval_brain = qtl_analysis(brain_phenotypes[brain_phenotypes['Phenotype'].str.contains(\"Morphine\")], brain_genotypes, file_prefix=\"brain_morphine_\")\n",
    "logpval_liver = qtl_analysis(liver_phenotypes[liver_phenotypes['Phenotype'].str.contains(\"Morphine\")], liver_genotypes, file_prefix=\"liver_morphine_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "assoc_qtl_l, ntests_qtl_l = get_associations(logpval_brain) # by bonfferoni corrections\n",
    "assoc_qtl_b, ntests_qtl_b = get_associations(logpval_liver) # by bonfferoni corrections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>gene</th>\n",
       "      <th>p-value</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>minus_log_p-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SNP, chromosome, position, gene, p-value, phenotype, minus_log_p-value]\n",
       "Index: []"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assoc_qtl_b"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6. Causality Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selected triplet for causality analysis:\n",
      "1. Locus (L): rs13477308\n",
      "2. Gene expression (R): Cenpl\n",
      "3. Phenotype (C): Anxiety assay, restraint stress [15 min] + ethanol treated [1.8 g/kg i.p.] (RSE group), activity in closed quadrants using an elevated zero maze in 60 to 120-day-old males and females during last 5 min [n beam breaks]\n"
     ]
    }
   ],
   "source": [
    "# Choose a triplet L, R, C for the causality analysis\n",
    "locus_idx = 241      \n",
    "gene_exp_idx = 288 \n",
    "phenotype_idx = 1842\n",
    "\n",
    "locus = liver_genotypes.loc[[locus_idx]].replace({'B': 0, 'b': 0, 'D':1, 'H': np.nan, 'U': np.nan})[liver_common_bxd]\n",
    "gene_exp = liver_matrix.loc[[gene_exp_idx]][liver_common_bxd]\n",
    "phenotype = liver_phenotypes.loc[[phenotype_idx]][liver_common_bxd]\n",
    "\n",
    "print(\"Selected triplet for causality analysis:\\n1. Locus (L): %s\\n2. Gene expression (R): %s\\n3. Phenotype (C): %s\"\n",
    "     % (liver_genotypes.iloc[locus_idx]['Locus'],\n",
    "        liver_matrix.iloc[gene_exp_idx]['GENE_SYMBOL'],\n",
    "        liver_phenotypes.iloc[phenotype_idx]['Phenotype']))\n",
    "\n",
    "# Create input table \n",
    "input_table = pd.concat([locus, gene_exp, phenotype]) # concate data types\n",
    "input_table = input_table.T.reset_index()             # transpose\n",
    "input_table.columns = ['Individual', 'L', 'R', 'C']\n",
    "input_table = input_table.dropna()                    # ignore nulls\n",
    "input_table = input_table.sort_values(by='L')         # sort by lucus (genotype 0 or 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Individual</th>\n",
       "      <th>L</th>\n",
       "      <th>R</th>\n",
       "      <th>C</th>\n",
       "      <th>R|L_mean</th>\n",
       "      <th>R|L_var</th>\n",
       "      <th>C|R_mean</th>\n",
       "      <th>C|R_var</th>\n",
       "      <th>P(R|L)</th>\n",
       "      <th>P(C|R)</th>\n",
       "      <th>P(L) * P(C|R) * P(R|L)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BXD1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.7185</td>\n",
       "      <td>787.00</td>\n",
       "      <td>0.641457</td>\n",
       "      <td>0.346496</td>\n",
       "      <td>585.057</td>\n",
       "      <td>14697.1</td>\n",
       "      <td>0.671956</td>\n",
       "      <td>0.000822</td>\n",
       "      <td>0.000276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BXD85</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.2175</td>\n",
       "      <td>605.67</td>\n",
       "      <td>0.641457</td>\n",
       "      <td>0.346496</td>\n",
       "      <td>642.945</td>\n",
       "      <td>14697.1</td>\n",
       "      <td>0.419863</td>\n",
       "      <td>0.003139</td>\n",
       "      <td>0.000659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BXD8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8990</td>\n",
       "      <td>566.00</td>\n",
       "      <td>0.641457</td>\n",
       "      <td>0.346496</td>\n",
       "      <td>605.996</td>\n",
       "      <td>14697.1</td>\n",
       "      <td>0.615875</td>\n",
       "      <td>0.003116</td>\n",
       "      <td>0.000960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>BXD77</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6190</td>\n",
       "      <td>361.23</td>\n",
       "      <td>0.641457</td>\n",
       "      <td>0.346496</td>\n",
       "      <td>573.515</td>\n",
       "      <td>14697.1</td>\n",
       "      <td>0.677243</td>\n",
       "      <td>0.000710</td>\n",
       "      <td>0.000241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BXD73</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.2045</td>\n",
       "      <td>572.50</td>\n",
       "      <td>0.641457</td>\n",
       "      <td>0.346496</td>\n",
       "      <td>641.437</td>\n",
       "      <td>14697.1</td>\n",
       "      <td>0.428931</td>\n",
       "      <td>0.002799</td>\n",
       "      <td>0.000600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Individual    L       R       C  R|L_mean   R|L_var C|R_mean  C|R_var  \\\n",
       "0        BXD1  0.0  0.7185  787.00  0.641457  0.346496  585.057  14697.1   \n",
       "36      BXD85  0.0  1.2175  605.67  0.641457  0.346496  642.945  14697.1   \n",
       "35       BXD8  0.0  0.8990  566.00  0.641457  0.346496  605.996  14697.1   \n",
       "34      BXD77  0.0  0.6190  361.23  0.641457  0.346496  573.515  14697.1   \n",
       "33      BXD73  0.0  1.2045  572.50  0.641457  0.346496  641.437  14697.1   \n",
       "\n",
       "      P(R|L)    P(C|R)  P(L) * P(C|R) * P(R|L)  \n",
       "0   0.671956  0.000822                0.000276  \n",
       "36  0.419863  0.003139                0.000659  \n",
       "35  0.615875  0.003116                0.000960  \n",
       "34  0.677243  0.000710                0.000241  \n",
       "33  0.428931  0.002799                0.000600  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m1_df, m2_df, m3_df = generate_models_df(input_table)\n",
    "m1_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate The Models Likelihood and Likelihood Ratio:\n",
    "\n",
    "Likelihood = L(θ;M) = p(data|θm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best model:  1\n",
      "The Likelihood-Ratio is: 2.53\n"
     ]
    }
   ],
   "source": [
    "LR = get_LR(m1_df, m2_df, m3_df)\n",
    "print(\"The Likelihood-Ratio is: %.2f\" % LR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Statistical estimation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO: one sample t-test is good?\n",
    "TODO: should we correct the pval? (p/100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  1\n",
      "Best model:  1\n",
      "Best model:  2\n",
      "Best model:  1\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  1\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  1\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  1\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  1\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  1\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  1\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  1\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  1\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  3\n",
      "Best model:  2\n",
      "Best model:  2\n",
      "Best model:  1\n",
      "Best model:  3\n",
      "Best model:  1\n",
      "Best model:  2\n",
      "Our LR is significant with pval  1.3575225854130205e-11\n"
     ]
    }
   ],
   "source": [
    "n_tests = 100\n",
    "LR_dist = []\n",
    "for i in range(0,n_tests):\n",
    "    shuffled_df = get_shuffled_df(input_table.copy())\n",
    "    m1_df, m2_df, m3_df = generate_models_df(shuffled_df)\n",
    "    LR_i = get_LR(m1_df, m2_df, m3_df)\n",
    "    LR_dist.append(LR_i)\n",
    "    \n",
    "pval = stats.ttest_1samp(LR_dist, LR)[1]\n",
    "if pval <= 0.05:\n",
    "    print(\"Our LR is significant with pval \", pval)\n",
    "else:\n",
    "    print(\"Our LR is not significant \", pval)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
